### ISSUE TITLE
Can Persistent filter with a comparison on two fields in the same model?

### ISSUE THREAD

**Issue by ketzacoatl** [yesodweb/persistent#1309]
Hello,

Thank you for such a great library!

I believe I've run into a limitation in persistent, but maybe it's already possible to do this and I'm overlooking how to do it?

I have a model like the following:

```
Thing
  ...
  fooVal Int
  barVal Int
```

and I'd like to use `selectList` with a filter that compares those values such as:

```
selectThings = runDB $ selectList [ThingFooVal >. ThingBarVal] []
```

This fails with a type error like:

```
Couldn't match type 'EntityField Thing Int' with 'Int' arising from a use of ThingBarVal
```

Is this not possible in persistent directly, or is there a different way I should setup the query?

**Comment by parsonsmatt** [comment]
This isn't possible with `persistent`'s API right now. It may be feasible by hacking the `Filter` constructor directly.

You can do this with `esqueleto`:

```haskell
selectThings = runDB $ 
    select $ do
      thing <- from $ table @Thing
      where_ $ thing ^. ThingFooVal >. thing ^. ThingBarVal
      pure thing
```

**Comment by ketzacoatl** [comment]
Thank you for the confirmation @parsonsmatt, I appreciate it :)

Is this a difficult thing to add support for? I'd be happy to help if it's feasible.

I will see if I can figure out `esqueleto` here in the meantime

**Comment by parsonsmatt** [comment]
🤔 

The relevant type is here:

```haskell
module Database.Persist.Class.PersistEntity where 

...

-- | Filters which are available for 'select', 'updateWhere' and
-- 'deleteWhere'. Each filter constructor specifies the field being
-- filtered on, the type of comparison applied (equals, not equals, etc)
-- and the argument for the comparison.
--
-- Persistent users use combinators to create these.
--
-- Note that it's important to be careful about the 'PersistFilter' that
-- you are using, if you use this directly. For example, using the 'In'
-- 'PersistFilter' requires that you have an array- or list-shaped
-- 'EntityField'. It is possible to construct values using this that will
-- create malformed runtime values.
data Filter record = forall typ. PersistField typ => Filter
    { filterField  :: EntityField record typ
    , filterValue  :: FilterValue typ
    , filterFilter :: PersistFilter -- FIXME
    }
    | FilterAnd [Filter record] -- ^ convenient for internal use, not needed for the API
    | FilterOr  [Filter record]
    | BackendFilter
          (BackendSpecificFilter (PersistEntityBackend record) record)

-- | Value to filter with. Highly dependant on the type of filter used.
--
-- @since 2.10.0
data FilterValue typ where
  FilterValue  :: typ -> FilterValue typ
  FilterValues :: [typ] -> FilterValue typ
  UnsafeValue  :: forall a typ. PersistField a => a -> FilterValue typ
```

The `>.` operator to construct it is here:

```haskell
module Database.Persist where

...
(>.) :: forall v typ.  PersistField typ => EntityField v typ -> typ -> Filter v
f >. a  = Filter f (FilterValue a) Gt
```

We'd need to define a new operator - `(>.)` requires that the RHS be a Haskell value `typ` corresponding to the `EntityField record typ` that you provide. The `UnsafeValue` constructor is *almost* what we want - but there's no `PersistField` instance for `EntityField record typ`.

So, for a rather nasty hack, we can possibly write:

```haskell
instance PersistField (EntityField Thing typ) where
  toPersistValue = PersistLiteralUnescaped . go
    where
      go field = case field of
        ThingFooVal -> "foo_val"
        ThingBarVal -> "bar_val"
        -- TODO: refactor to use persistFieldDef
  fromPersistValue =
    error "this is not sensible"
```

which would let us write:

```haskell
(>.!) :: PersistField typ => EntityField v typ -> EntityField v typ -> Filter v
f >.! f' = Filter f (UnsafeValue f') Gt
```

This *may* work but it's a pretty gnarly hack. 

---

The smallest change to add support is a new constructor for `Filter` type.

```haskell
data Filter record where
  ...
  FilterField :: EntityField record typ -> FilterOp -> EntityField record typ -> Filter record
```

With this, we can write:

```haskell
(>.!) :: PersistField typ => EntityField v typ -> EntityField v typ -> Filter v
f0 (>.!) f1 = FilterField f0 Gt f1
```

**Comment by ketzacoatl** [comment]
I'm going to take a moment to digest that, and have a couple questions, so please excuse my lack of experience with the haskell topics here.

> The smallest change to add support is a new constructor for Filter type.

It seems that could be tested on its own in a local project directly, rather than in the persistent library as a fork, am I following?

> This may work but it's a pretty gnarly hack.

Given that this is possible in `esqueleto`, and adding gnarly hacks isn't really the direction to go, should persistent skip this? Or if the smaller change works, should that be added in parallel, as a new `FooFilter`?

**Comment by parsonsmatt** [comment]
Yeah, as an end-user, I'd suggest learning `esqueleto` for anything that `persistent` doesn't give you. You'll want it anyway for `JOIN`s.

The "add a constructor for `Filter`" would require patching `persistent` as a library,  so you'd need to fork `persistent` to do it.

Adding the gnarly hack (the `PersistField (EntityField r t)` instance would be possible in your local project, but not recommended.
